tsl/src/compression/compression.c

```diff
@@ -2342,35 +2393,50 @@ decompress_batches(Chunk *ch, ScanKeyData *scankeys, int num_scankeys, Bitmapset
 void
 decompress_batches_for_update_delete(List *chunks, List *predicates)
 {
-	List *segmentby_columns = NIL;
-	List *segmentby_columns_value = NIL;
+	List *filter_columns = NIL;
+	List *filter_ops = NIL;
+	List *filter_columns_value = NIL;
 	List *is_null_check = NIL;
 	List *is_null = NIL;
 	ListCell *ch = NULL;
+
+	Relation chunk_rel;
+	Relation comp_chunk_rel;
+	Chunk *chunk, *comp_chunk;
+	RowDecompressor decompressor;
+
 	if (predicates)
 		fill_predicate_context(linitial(chunks),
 							   predicates,
-							   &segmentby_columns,
-							   &segmentby_columns_value,
+							   &filter_columns,
+							   &filter_ops,
+							   &filter_columns_value,
 							   &is_null_check,
 							   &is_null);
 	foreach (ch, chunks)
 	{
+		chunk = (Chunk *) lfirst(ch);
 		bool chunk_status_changed = false;
 		ScanKeyData *scankeys = NULL;
 		Bitmapset *null_columns = NULL;
 		int num_scankeys = 0;
 
-		if (segmentby_columns)
+		chunk_rel = table_open(chunk->table_id, RowExclusiveLock);
+		comp_chunk = ts_chunk_get_by_id(chunk->fd.compressed_chunk_id, true);
+		comp_chunk_rel = table_open(comp_chunk->table_id, RowExclusiveLock);
+		decompressor = build_decompressor(comp_chunk_rel, chunk_rel);
```

Comment:
Just an observation, build_decompressor has call to create_per_compressed_column() which seems to be unwanted in this case for each chunk.

Reply:
Maybe we can optimize it away in a future change, for now I would leave it as is since we should also measure the impact of the optimization.

Seems worthy of a separate PR.

Reply:
Ok.